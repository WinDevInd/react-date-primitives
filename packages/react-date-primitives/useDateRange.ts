import { useState, useRef } from 'react';

import { DayOfMonth, isDayBefore, isDayAfter, isSameDay } from './utils';

export interface DayOfRangeMonth extends DayOfMonth {
    inRange: boolean;
    selected: boolean;
    isStart: boolean;
    isEnd: boolean;
}

export interface UseDateRange {
    handleDayClick(day: Date): void;
    handleDayHover(day: Date): void;
    startDate?: Date;
    endDate?: Date;
    processMonth(month: DayOfMonth[][]): DayOfRangeMonth[][];
}

export function useDateRange(initialStartDate?: Date, initialEndDate?: Date): UseDateRange {
    const [startDate, setStartDate] = useState(initialStartDate);
    const [endDate, setEndDate] = useState(initialEndDate);
    const selection = useRef(false);

    function handleDayClick(day: Date) {
        if (startDate && selection.current && !isDayBefore(day, startDate)) {
            setStartDate(startDate);
            setEndDate(day);
            selection.current = false;
        } else {
            setStartDate(day);
            setEndDate(undefined);
            selection.current = true;
        }
    }

    function handleDayHover(day: Date) {
        if (startDate && selection.current && !isDayBefore(day, startDate)) {
            setEndDate(day);
        }
    }

    function processMonth(month: DayOfMonth[][]): DayOfRangeMonth[][] {
        return month.map<DayOfRangeMonth[]>((week) =>
            week.map<DayOfRangeMonth>((day) => {
                const isStart = Boolean(startDate && isSameDay(day.date, startDate));
                const isEnd = Boolean(endDate && isSameDay(day.date, endDate));

                return {
                    ...day,
                    selected: isStart || isEnd,
                    inRange:
                        isStart ||
                        isEnd ||
                        Boolean(
                            startDate &&
                                isDayAfter(day.date, startDate) &&
                                endDate &&
                                isDayBefore(day.date, endDate)
                        ),
                    isStart,
                    isEnd
                };
            })
        );
    }

    return { handleDayClick, handleDayHover, startDate, endDate, processMonth };
}
