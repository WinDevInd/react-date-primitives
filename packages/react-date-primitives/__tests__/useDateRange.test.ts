import { renderHook, act } from 'react-hooks-testing-library';
import { useDateRange } from '../useDateRange';

import { isSameDay, getDaysOfMonth } from '../utils';

describe('useDateRange hook tests', () => {
    test('selection (handleDayClick) works for endDate > startDate', () => {
        const date1 = new Date(2019, 0, 1, 0, 0, 0, 0);
        const date2 = new Date(2019, 0, 9, 0, 0, 0, 0);
        const { result } = renderHook(() => useDateRange());

        expect(result.current.startDate).not.toBeDefined();
        expect(result.current.endDate).not.toBeDefined();

        // set a startDate
        act(() => {
            result.current.handleDayClick(date1);
        });
        expect(isSameDay(result.current.startDate!, date1)).toBe(true);
        expect(result.current.endDate).not.toBeDefined();

        // set an endDate > startDate
        act(() => {
            result.current.handleDayClick(date2);
        });
        expect(isSameDay(result.current.startDate!, date1)).toBe(true);
        expect(isSameDay(result.current.endDate!, date2)).toBe(true);

        const date3 = new Date(2019, 2, 1, 0, 0, 0, 0);
        const date4 = new Date(2019, 2, 9, 0, 0, 0, 0);

        // set a startDate
        act(() => {
            result.current.handleDayClick(date3);
        });
        expect(isSameDay(result.current.startDate!, date3)).toBe(true);
        expect(result.current.endDate).not.toBeDefined();

        // set an endDate > startDate
        act(() => {
            result.current.handleDayClick(date4);
        });
        expect(isSameDay(result.current.startDate!, date3)).toBe(true);
        expect(isSameDay(result.current.endDate!, date4)).toBe(true);
    });

    test('selection (handleDayClick) resets for endDate < startDate', () => {
        const date1 = new Date(2019, 0, 9, 0, 0, 0, 0);
        const date2 = new Date(2019, 0, 1, 0, 0, 0, 0);
        const { result } = renderHook(() => useDateRange());

        expect(result.current.startDate).not.toBeDefined();
        expect(result.current.endDate).not.toBeDefined();

        // set a startDate
        act(() => {
            result.current.handleDayClick(date1);
        });
        expect(isSameDay(result.current.startDate!, date1)).toBe(true);
        expect(result.current.endDate).not.toBeDefined();

        // set an endDate < startDate
        act(() => {
            result.current.handleDayClick(date2);
        });
        expect(isSameDay(result.current.startDate!, date2)).toBe(true);
        expect(result.current.endDate).not.toBeDefined();
    });

    test('handleDayOver sets endDate temporarily', () => {
        const date = new Date(2019, 0, 15, 0, 0, 0, 0);
        const prevDate = new Date(2019, 0, 10, 0, 0, 0, 0);
        const nextDate1 = new Date(2019, 0, 20, 0, 0, 0, 0);
        const nextDate2 = new Date(2019, 0, 25, 0, 0, 0, 0);

        const { result } = renderHook(() => useDateRange());

        expect(result.current.startDate).not.toBeDefined();
        expect(result.current.endDate).not.toBeDefined();

        // set a startDate
        act(() => {
            result.current.handleDayClick(date);
        });
        expect(isSameDay(result.current.startDate!, date)).toBe(true);
        expect(result.current.endDate).not.toBeDefined();

        // hover over a date < startDate
        act(() => {
            result.current.handleDayHover(prevDate);
        });
        expect(isSameDay(result.current.startDate!, date)).toBe(true);
        expect(result.current.endDate).not.toBeDefined();

        // hover over a date > startDate
        act(() => {
            result.current.handleDayHover(nextDate1);
        });
        expect(isSameDay(result.current.startDate!, date)).toBe(true);
        expect(isSameDay(result.current.endDate!, nextDate1)).toBe(true);

        // hover over a date > startDate
        act(() => {
            result.current.handleDayHover(nextDate2);
        });
        expect(isSameDay(result.current.startDate!, date)).toBe(true);
        expect(isSameDay(result.current.endDate!, nextDate2)).toBe(true);
    });

    test('processMonth adds extra info to days', () => {
        const date1 = new Date(2019, 0, 1, 0, 0, 0, 0);
        const date2 = new Date(2019, 0, 9, 0, 0, 0, 0);
        const { result } = renderHook(() => useDateRange());
        const days = getDaysOfMonth(date1);

        // set a startDate
        act(() => {
            result.current.handleDayClick(date1);
        });

        // set an endDate > startDate
        act(() => {
            result.current.handleDayClick(date2);
        });

        const expected = result.current
            .processMonth(days)
            .map((week) => week.map((day) => ({ ...day, date: day.date.toDateString() })));

        expect(expected).toMatchSnapshot();
    });
});
