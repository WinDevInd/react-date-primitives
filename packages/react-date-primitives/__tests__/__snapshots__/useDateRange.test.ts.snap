// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useDateRange hook tests processMonth adds extra info to days 1`] = `
Array [
  Array [
    Object {
      "date": "Sun Dec 30 2018",
      "inCurrentMonth": false,
      "inRange": false,
      "isEnd": false,
      "isStart": false,
      "selected": false,
    },
    Object {
      "date": "Mon Dec 31 2018",
      "inCurrentMonth": false,
      "inRange": false,
      "isEnd": false,
      "isStart": false,
      "selected": false,
    },
    Object {
      "date": "Tue Jan 01 2019",
      "inCurrentMonth": true,
      "inRange": true,
      "isEnd": false,
      "isStart": true,
      "selected": true,
    },
    Object {
      "date": "Wed Jan 02 2019",
      "inCurrentMonth": true,
      "inRange": true,
      "isEnd": false,
      "isStart": false,
      "selected": false,
    },
    Object {
      "date": "Thu Jan 03 2019",
      "inCurrentMonth": true,
      "inRange": true,
      "isEnd": false,
      "isStart": false,
      "selected": false,
    },
    Object {
      "date": "Fri Jan 04 2019",
      "inCurrentMonth": true,
      "inRange": true,
      "isEnd": false,
      "isStart": false,
      "selected": false,
    },
    Object {
      "date": "Sat Jan 05 2019",
      "inCurrentMonth": true,
      "inRange": true,
      "isEnd": false,
      "isStart": false,
      "selected": false,
    },
  ],
  Array [
    Object {
      "date": "Sun Jan 06 2019",
      "inCurrentMonth": true,
      "inRange": true,
      "isEnd": false,
      "isStart": false,
      "selected": false,
    },
    Object {
      "date": "Mon Jan 07 2019",
      "inCurrentMonth": true,
      "inRange": true,
      "isEnd": false,
      "isStart": false,
      "selected": false,
    },
    Object {
      "date": "Tue Jan 08 2019",
      "inCurrentMonth": true,
      "inRange": true,
      "isEnd": false,
      "isStart": false,
      "selected": false,
    },
    Object {
      "date": "Wed Jan 09 2019",
      "inCurrentMonth": true,
      "inRange": true,
      "isEnd": true,
      "isStart": false,
      "selected": true,
    },
    Object {
      "date": "Thu Jan 10 2019",
      "inCurrentMonth": true,
      "inRange": false,
      "isEnd": false,
      "isStart": false,
      "selected": false,
    },
    Object {
      "date": "Fri Jan 11 2019",
      "inCurrentMonth": true,
      "inRange": false,
      "isEnd": false,
      "isStart": false,
      "selected": false,
    },
    Object {
      "date": "Sat Jan 12 2019",
      "inCurrentMonth": true,
      "inRange": false,
      "isEnd": false,
      "isStart": false,
      "selected": false,
    },
  ],
  Array [
    Object {
      "date": "Sun Jan 13 2019",
      "inCurrentMonth": true,
      "inRange": false,
      "isEnd": false,
      "isStart": false,
      "selected": false,
    },
    Object {
      "date": "Mon Jan 14 2019",
      "inCurrentMonth": true,
      "inRange": false,
      "isEnd": false,
      "isStart": false,
      "selected": false,
    },
    Object {
      "date": "Tue Jan 15 2019",
      "inCurrentMonth": true,
      "inRange": false,
      "isEnd": false,
      "isStart": false,
      "selected": false,
    },
    Object {
      "date": "Wed Jan 16 2019",
      "inCurrentMonth": true,
      "inRange": false,
      "isEnd": false,
      "isStart": false,
      "selected": false,
    },
    Object {
      "date": "Thu Jan 17 2019",
      "inCurrentMonth": true,
      "inRange": false,
      "isEnd": false,
      "isStart": false,
      "selected": false,
    },
    Object {
      "date": "Fri Jan 18 2019",
      "inCurrentMonth": true,
      "inRange": false,
      "isEnd": false,
      "isStart": false,
      "selected": false,
    },
    Object {
      "date": "Sat Jan 19 2019",
      "inCurrentMonth": true,
      "inRange": false,
      "isEnd": false,
      "isStart": false,
      "selected": false,
    },
  ],
  Array [
    Object {
      "date": "Sun Jan 20 2019",
      "inCurrentMonth": true,
      "inRange": false,
      "isEnd": false,
      "isStart": false,
      "selected": false,
    },
    Object {
      "date": "Mon Jan 21 2019",
      "inCurrentMonth": true,
      "inRange": false,
      "isEnd": false,
      "isStart": false,
      "selected": false,
    },
    Object {
      "date": "Tue Jan 22 2019",
      "inCurrentMonth": true,
      "inRange": false,
      "isEnd": false,
      "isStart": false,
      "selected": false,
    },
    Object {
      "date": "Wed Jan 23 2019",
      "inCurrentMonth": true,
      "inRange": false,
      "isEnd": false,
      "isStart": false,
      "selected": false,
    },
    Object {
      "date": "Thu Jan 24 2019",
      "inCurrentMonth": true,
      "inRange": false,
      "isEnd": false,
      "isStart": false,
      "selected": false,
    },
    Object {
      "date": "Fri Jan 25 2019",
      "inCurrentMonth": true,
      "inRange": false,
      "isEnd": false,
      "isStart": false,
      "selected": false,
    },
    Object {
      "date": "Sat Jan 26 2019",
      "inCurrentMonth": true,
      "inRange": false,
      "isEnd": false,
      "isStart": false,
      "selected": false,
    },
  ],
  Array [
    Object {
      "date": "Sun Jan 27 2019",
      "inCurrentMonth": true,
      "inRange": false,
      "isEnd": false,
      "isStart": false,
      "selected": false,
    },
    Object {
      "date": "Mon Jan 28 2019",
      "inCurrentMonth": true,
      "inRange": false,
      "isEnd": false,
      "isStart": false,
      "selected": false,
    },
    Object {
      "date": "Tue Jan 29 2019",
      "inCurrentMonth": true,
      "inRange": false,
      "isEnd": false,
      "isStart": false,
      "selected": false,
    },
    Object {
      "date": "Wed Jan 30 2019",
      "inCurrentMonth": true,
      "inRange": false,
      "isEnd": false,
      "isStart": false,
      "selected": false,
    },
    Object {
      "date": "Thu Jan 31 2019",
      "inCurrentMonth": true,
      "inRange": false,
      "isEnd": false,
      "isStart": false,
      "selected": false,
    },
    Object {
      "date": "Fri Feb 01 2019",
      "inCurrentMonth": false,
      "inRange": false,
      "isEnd": false,
      "isStart": false,
      "selected": false,
    },
    Object {
      "date": "Sat Feb 02 2019",
      "inCurrentMonth": false,
      "inRange": false,
      "isEnd": false,
      "isStart": false,
      "selected": false,
    },
  ],
  Array [
    Object {
      "date": "Sun Feb 03 2019",
      "inCurrentMonth": false,
      "inRange": false,
      "isEnd": false,
      "isStart": false,
      "selected": false,
    },
    Object {
      "date": "Mon Feb 04 2019",
      "inCurrentMonth": false,
      "inRange": false,
      "isEnd": false,
      "isStart": false,
      "selected": false,
    },
    Object {
      "date": "Tue Feb 05 2019",
      "inCurrentMonth": false,
      "inRange": false,
      "isEnd": false,
      "isStart": false,
      "selected": false,
    },
    Object {
      "date": "Wed Feb 06 2019",
      "inCurrentMonth": false,
      "inRange": false,
      "isEnd": false,
      "isStart": false,
      "selected": false,
    },
    Object {
      "date": "Thu Feb 07 2019",
      "inCurrentMonth": false,
      "inRange": false,
      "isEnd": false,
      "isStart": false,
      "selected": false,
    },
    Object {
      "date": "Fri Feb 08 2019",
      "inCurrentMonth": false,
      "inRange": false,
      "isEnd": false,
      "isStart": false,
      "selected": false,
    },
    Object {
      "date": "Sat Feb 09 2019",
      "inCurrentMonth": false,
      "inRange": false,
      "isEnd": false,
      "isStart": false,
      "selected": false,
    },
  ],
]
`;
